bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/PacketListener.java

```diff
@@ -0,0 +1,145 @@
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.smaenergymeter.internal.handler.EnergyMeter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link PacketListener} class is responsible for communication with the SMA devices.
+ * It handles udp/multicast traffic and broadcast received data to subsequent payload handlers.
+ *
+ * @author Łukasz Dywicki - Initial contribution
+ */
+
+@NonNullByDefault
+public class PacketListener {
+
+    private final DefaultPacketListenerRegistry registry;
+    private final List<PayloadHandler> handlers = new CopyOnWriteArrayList<>();
+
+    private String multicastGroup;
+    private int port;
+
+    public static final String DEFAULT_MCAST_GRP = "239.12.255.254";
+    public static final int DEFAULT_MCAST_PORT = 9522;
+
+    private @Nullable MulticastSocket socket;
+    private @Nullable ScheduledFuture<?> future;
+
+    public PacketListener(DefaultPacketListenerRegistry registry, String multicastGroup, int port) {
+        this.registry = registry;
+        this.multicastGroup = multicastGroup;
+        this.port = port;
+    }
+
+    public void addPayloadHandler(PayloadHandler handler) {
+        handlers.add(handler);
+    }
+
+    public void removePayloadHandler(PayloadHandler handler) {
+        handlers.remove(handler);
+
+        if (handlers.isEmpty()) {
+            registry.close(multicastGroup, port);
+        }
+    }
+
+    public boolean isOpen() {
+        MulticastSocket socket = this.socket;
+        return socket != null && socket.isConnected();
+    }
+
+    public void open(int intervalSec) throws IOException {
+        if (isOpen()) {
+            // no need to bind socket second time
+            return;
+        }
+        MulticastSocket socket = new MulticastSocket(port);
+        socket.setSoTimeout(5000);
+        InetAddress address = InetAddress.getByName(multicastGroup);
+        socket.joinGroup(address);
+
+        future = registry.addTask(new ReceivingTask(socket, multicastGroup + ":" + port, handlers), intervalSec);
+        this.socket = socket;
+    }
+
+    void close() throws IOException {
+        ScheduledFuture<?> future = this.future;
+        if (future != null) {
+            future.cancel(true);
+            this.future = null;
+        }
+
+        InetAddress address = InetAddress.getByName(multicastGroup);
+        MulticastSocket socket = this.socket;
+        if (socket != null) {
+            socket.leaveGroup(address);
+            socket.close();
+            this.socket = null;
+        }
+    }
+
+    public void request() {
+        registry.execute(new ReceivingTask(socket, multicastGroup + ":" + port, handlers));
+    }
+
+    static class ReceivingTask implements Runnable {
+        private final Logger logger = LoggerFactory.getLogger(ReceivingTask.class);
+        private final @Nullable DatagramSocket socket;
```

Comment:
Maybe it is better to have it there not nullable and to check if the socket parameter is not null when instantiating the class ?

Reply:
Don't think this matters much, but done.